text/0000-dynamic-attributes.md

```diff
@@ -0,0 +1,189 @@
+`lwc:spread` though, it can be present in one render and not be present in the next.
+
+- attrs/listeners are added/removed based on their presence/absence in `lwc:spread`.
+- properties can't be "removed" if they are not present in subsequent renders. So, we set them to `undefined`. New properties are
+set as usual.
+
+An interesting side effect of the above behavior can be observed when properties have default values:
+
+```html
+// x/app.html
+<template>
+    <x-child lwc:spread={props}></x-child>
+</template>
+```
+```js
+export default class Child extends LightningElement {
+    @api name = "lwc";
+}
+
+export default class App extends LightningElement {
+    props = {};
+    renderedCallback() {
+        // in initial render props is {} and `name` in child is 'lwc'
+        this.props = {name: "aura"}; // `name` in child is "aura";
+        this.props = {}; // `name` in child is `undefined`;
+    }
+}
+
+```
+Customers should be educated on the above behavior.
+
+### Reactivity
+There's no precendent in LWC for change in the set of attributes/properties/listeners. An element can't have a property in one
+render and not have it in another render. LWC Engine actually throws when it detects such a change. In the case of `lwc:spread`,
+we'll have to relax this validation.
+
+In every render, `lwc:spread` will merge with the older value of `lwc:spread`. Any keys not present in the new `lwc:spread` but
+present in the old `lwc:spread` will be set to `null` explicitly in `lwc:spread`. With this merging logic, all the new
+attributes/listeners will be set and missing attributes/listeners will be removed. Missing properties will be explicitly set to `null`. This results in a consistent, logical behavior for users. Please see the following example.
+
+### Example
+```html
+// x/test.html
+<template>
+    <x-child lwc:spread={props}></x-child>
+    <button onclick={toggle}>Toggle</button>
+</template>
+```
+
+```js
+import { LightningElement } from 'lwc';
+
+export default class Test extends LightningElement {
+    props = {
+        attrA: 'a',
+        attrB: 'b',
+
+        propA: 'a',
+        propB: 'b',
+
+        listenerA() { },
+        listenerB() { },
+    }
+    toggle() {
+        this.props = {
+            attrA: 'a', // remains same
+            attrC: 'c', // new attribute added using setAttribute
+            // attrB is removed using `removeAttribute`
+
+            propA: 'a', // remains same
+            propC: 'c', // new property assigned using elm.propC = 'c'
+            // propB is "removed" using elm.PropB = null;
+
+            listenerA() { }, // remains same
+            listenerC() { }, // new listener added using `addEventListener`
+            // listenerB is removed using `removeEventListener`
+        }
+    }
+}
+```
+
+### Casing
+LWC Template Compiler converts kebab-case attrs/props to camelCase to match the Javascript counterparts. Since props `lwc:spread`
```

Comment:
The first sentence is half true... because we only do that for CE, built-ins are using attrs most of the time. I don't think it is relevant... but maybe just saying that they must be camelCase for properties, and no attributes, should be sufficient.

Reply:
I don't think we should talk here about what we do... but the semantics of the new feature instead. What we do now can change tomorrow, but the semantics of this feature will remain intact. Basically, just saying that the value on the spread needs to use prop names in a camelCase. Maybe also listing a couple of exceptions to the rule to ilustrate the semantics.